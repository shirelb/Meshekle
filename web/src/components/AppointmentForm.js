import React, {Component} from 'react';
import {Dropdown, Form, Message, TextArea} from 'semantic-ui-react';
import {post} from "axios";
import moment from 'moment';
import Datetime from 'react-datetime';
import 'react-datetime/css/react-datetime.css';
import helpers from "../shared/helpers";
import './styles.css';


const subjectOptions = [
    {key: 'f', text: 'פן', value: 'פן'},
    {key: 'hd', text: 'צבע', value: 'צבע'},
    {key: 'hc', text: 'תספורת', value: 'תספורת'},
];

let userOptions = {};

helpers.getUsers()
    .then(users => {
        console.log('users ', users);
        userOptions = users.map(item => ({
            key: item.userId,
            text: item.fullname,
            value: item.fullname
        }));
    });


class AppointmentForm extends Component {

    constructor(props) {
        super(props);

        const {slotInfo, appointment = {}} = props;

        this.state = {
            appointment: {
                date: moment(slotInfo.start),
                startTime: '',
                endTime: '',
                subject: [],
                clientName: '',
            },
            formError: false,
            formComplete: false,
        };

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        const {appointment} = nextProps;

        this.setState({appointment});
    }

    handleSubmit(e) {
        e.preventDefault();
        const {appointment} = this.state;
        const {handleSubmit} = this.props;

        if (appointment.clientName !== '' &&
            appointment.subject.length > 0 &&
            appointment.date !== '' &&
            appointment.startTime !== '' &&
            appointment.endTime !== '') {
            this.setState({formComplete: true});
            let updateAppointment = appointment;
            updateAppointment.clientId = (userOptions.filter(user => user.value === appointment.clientName))[0].key;
            this.setState({appointment: updateAppointment});

            handleSubmit(updateAppointment);
            this.setState({appointment: {}});
        } else {
            this.setState({formError: true});
        }
    }

    handleChange(e, {name, value}) {
        const {appointment} = this.state;

        this.setState({formError: false, formComplete: false});
        this.setState({appointment: {...appointment, [name]: value}});
    }

    handleClear = (e) => {
        e.preventDefault();
        this.setState({
            appointment: {
                date: '',
                startTime: '',
                endTime: '',
                subject: [],
                clientName: '',
                formError: false,
                formComplete: false,
            },
        });
    };

    onChangeDate = date => {
        let updateAppointment = this.state.appointment;
        updateAppointment.date = date;
        this.setState({appointment: updateAppointment})
    };

    onChangeTime = (time, isStart) => {
        let updateAppointment = this.state.appointment;
        isStart ?
            updateAppointment.startTime = time
            :
            updateAppointment.endTime = time;
        this.setState({appointment: updateAppointment});
    };

    handleSearchChange = (e, {searchQuery}) => {
        let updateAppointment = this.state.appointment;
        updateAppointment.subject = searchQuery;
        this.setState({appointment: updateAppointment})
    };


    render() {
        const {appointment, formError, formComplete} = this.state;
        const {handleCancel, submitText = 'קבע'} = this.props;

        return (
            <Form onSubmit={this.handleSubmit} error={formError}>
                <Form.Field
                    control={Dropdown}
                    label='שם לקוח'
                    placeholder='שם לקוח'
                    fluid
                    search
                    selection
                    autoComplete='on'
                    options={userOptions}
                    value={appointment.clientName}
                    onChange={this.handleChange}
                    name='clientName'
                    required
                    noResultsMessage='לא נמצאו התאמות'
                    // width='10'
                />
                <Form.Field
                    control={Dropdown}
                    label='נושא'
                    placeholder='נושא'
                    fluid
                    multiple
                    selection
                    options={subjectOptions}
                    value={appointment.subject}
                    onChange={this.handleChange}
                    name='subject'
                    required
                    noResultsMessage='לא נמצאו התאמות'
                    onSearchChange={this.handleSearchChange}
                    // width='10'
                />
                <Form.Field
                    control={TextArea}
                    label='הערות'
                    value={appointment.remarks}
                    name="remarks"
                    onChange={this.handleChange}
                    // width='10'
                />

                <Form.Group widths='equal'>
                    <Form.Field
                        as={Datetime}
                        label='תאריך'
                        value={appointment.date}
                        locale={'he'}
                        timeFormat={false}
                        install
                        // name="date"
                        onChange={this.onChangeDate.bind(this)}
                        required
                        // width='10'
                    />
                    <Form.Field
                        as={Datetime}
                        label='שעת התחלה'
                        value={appointment.startTime}
                        locale={'he'}
                        dateFormat={false}
                        install
                        // name="startTime"
                        onChange={time => this.onChangeTime(time, true)}
                        required
                        // width='10'
                    />
                    <Form.Field
                        as={Datetime}
                        label='שעת סיום'
                        value={appointment.endTime}
                        locale={'he'}
                        dateFormat={false}
                        install
                        // name="endTime"
                        onChange={time => this.onChangeTime(time, false)}
                        // onChange={this.handleChange}
                        required
                        // width='10'
                    />
                </Form.Group>

                {formError ?
                    <Message
                        error
                        header='פרטי תור חסרים'
                        content='נא להשלים את השדות החסרים'
                    />
                    : null
                }
                {formComplete ?
                    <Message success header='פרטי תור הושלמו' content="התור נקבע בהצלחה"/>
                    : null
                }

                <Form.Group>
                    <Form.Button type="submit">{submitText}</Form.Button>
                    <Form.Button onClick={handleCancel}>בטל</Form.Button>
                    <Form.Button onClick={this.handleClear}>נקה הכל</Form.Button>
                </Form.Group>

            </Form>
        )
    }
}

export default AppointmentForm