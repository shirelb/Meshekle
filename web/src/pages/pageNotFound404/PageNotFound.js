import React from 'react';
import {Grid, Header, Image} from 'semantic-ui-react';

const PageNotFound = () => (
    <Grid stretched textAlign={"center"}>
        <Grid.Column stretched textAlign={"center"}>
            <Image centered src={require('../../images/logo1.png')} size={'large'}/>
            <Header size={"huge"} textAlign={"center"} icon style={{marginTop: "-10px"}}>
                משקל'ה
                <Header.Subheader>יש משק ויש משקל'ה</Header.Subheader>
            </Header>

            <div style={{marginTop: 20, textAlign: "center"}}>
                <Header as={'h1'}> 404 </Header>
                <Header as={'h2'}> Page not found </Header>
            </div>
        </Grid.Column>
    </Grid>
);

export default PageNotFound;