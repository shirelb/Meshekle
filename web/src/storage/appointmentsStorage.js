import axios from "axios";
import {SERVER_URL, WEB_SOCKET} from "../shared/constants";
import moment from "moment";


var getAppointmentByAppointmentID = (serviceProviderId, appointmentId, headers) => {
    return axios.get(`${SERVER_URL}/api/appointments/serviceProvider/serviceProviderId/${serviceProviderId}`,
        {
            headers: headers,
            params: {
                appointmentId: appointmentId,
            }
        }
    )
        .then((response) => {
            let appointment = response.data[0];
            // console.log('getAppointmentByAppointmentID ', appointmentId, ' ', appointment);
            return appointment;
        })
        .catch((error) => {
            console.log('getAppointmentByAppointmentID ', appointmentId, ' ', error);
            return error;
        });
};

var getAppointmentRequestByAppointmentRequestID = (serviceProviderId, appointmentRequestId, headers) => {
    return axios.get(`${SERVER_URL}/api/appointmentRequests/serviceProvider/serviceProviderId/${serviceProviderId}`,
        {
            headers: headers,
            params: {
                appointmentRequestId: appointmentRequestId,
            }
        }
    )
        .then((response) => {
            let appointmentRequest = response.data[0];
            // console.log('getAppointmentByAppointmentID ', appointmentRequestId, ' ', appointmentRequest);
            return appointmentRequest;
        })
        .catch((error) => {
            console.log('getAppointmentByAppointmentID ', appointmentRequestId, ' ', error);
            return error;
        });
};

var getServiceProviderAppointmentRequests = (serviceProviderId, headers) => {
    return axios.get(`${SERVER_URL}/api/appointmentRequests/serviceProvider/serviceProviderId/${serviceProviderId}`,
        {
            headers: headers,
            params: {
                status: 'requested'
            },
        }
    )
        .then((response) => {
            return response;
        })
        .catch((error) => {
            console.log('get service provider appointment requests error ', error);
            return error;
        });
};

var getServiceProviderAppointments = (serviceProviderId, headers) => {
    return axios.get(`${SERVER_URL}/api/appointments/serviceProvider/serviceProviderId/${serviceProviderId}`,
        {
            headers: headers,
            params: {
                status: 'set'
            },
        }
    )
        .then((response) => {
            return response;
        })
        .catch((error) => {
            console.log('get service provider appointment requests error ', error);
            return error;
        });
};

var setAppointment = (appointment, serviceProviderId, roles, headers) => {
    return axios.post(`${SERVER_URL}/api/appointments/serviceProvider/set`,
        {
            userId: appointment.clientId,
            serviceProviderId: serviceProviderId,
            role: appointment.role,
            date: moment.isMoment(appointment.date) ? appointment.date.format('YYYY-MM-DD') : appointment.date,
            startHour: moment.isMoment(appointment.startTime) ? appointment.startTime.format("HH:mm") : appointment.startTime,
            endHour: moment.isMoment(appointment.endTime) ? appointment.endTime.format("HH:mm") : appointment.endTime,
            notes: appointment.remarks ? appointment.remarks : '',
            subject: JSON.stringify(appointment.subject)
        },
        {
            headers: headers
        }
    )
        .then((response) => {
            WEB_SOCKET.emit('serviceProviderPostAppointment', {
                userId: appointment.clientId,
            });

            return response;
        })
        .catch((error) => {
            console.log('add appointment ', error);
            return error;
        });
};

var updateAppointment = (event, headers) => {
    return axios.put(`${SERVER_URL}/api/appointments/serviceProvider/update/appointmentId/${event.appointmentId}`,
        {
            startDateAndTime: moment(event.date + ' ' + event.startTime).toDate(),
            endDateAndTime: moment(event.date + ' ' + event.endTime).toDate(),
            remarks: event.remarks,

            subject: JSON.stringify(event.subject),
            clientId: event.clientId,
            role: event.role,
        },
        {
            headers: headers,
        }
    )
        .then((response) => {
            WEB_SOCKET.emit('serviceProviderUpdateAppointment', {
                userId: event.clientId
            });

            return response;
        })
        .catch((error) => {
            console.log('update appointment error ', error);
            return error;
        });
};

var cancelAppointmentById = (appointment, headers) => {
    return axios.put(`${SERVER_URL}/api/appointments/serviceProvider/cancel/appointmentId/${appointment.appointmentId}`,
        {},
        {
            headers: headers
        }
    )
        .then((response) => {
            WEB_SOCKET.emit('serviceProviderCancelAppointment', {
                userId: appointment.AppointmentDetail.clientId,
            });

            return response
        })
        .catch((error) => {
            console.log('cancel appointment error ', error);
            return error;
        });
};

var rejectAppointmentRequestById = (appointmentRequest, headers) => {
    return axios.put(`${SERVER_URL}/api/appointmentRequests/serviceProvider/update/status/appointmentRequestId/${appointmentRequest.requestId}`,
        {},
        {
            headers: headers,
            params: {
                status: 'rejected'
            },
        }
    )
        .then((response) => {
            WEB_SOCKET.emit('serviceProviderRejectAppointmentRequest', {
                userId: appointmentRequest.AppointmentDetail.clientId,
            });

            return response
        })
        .catch((error) => {
            console.log('reject appointment request error ', error);
            return error;
        });
};

var approveAppointmentRequestById = (appointmentRequest, headers) => {
    return axios.put(`${SERVER_URL}/api/appointmentRequests/serviceProvider/update/status/appointmentRequestId/${appointmentRequest.requestId}`,
        {},
        {
            headers: headers,
            params: {
                status: 'approved'
            },
        }
    )
        .then((response) => {
            WEB_SOCKET.emit('serviceProviderApproveAppointmentRequest', {
                userId: appointmentRequest.AppointmentDetail.clientId,
            });

            return response
        })
        .catch((error) => {
            console.log('approve appointment request error ', error);
            return error;
        });
};


export default {
    getAppointmentByAppointmentID,
    getAppointmentRequestByAppointmentRequestID,
    setAppointment,
    cancelAppointmentById,
    rejectAppointmentRequestById,
    approveAppointmentRequestById,
    getServiceProviderAppointmentRequests,
    getServiceProviderAppointments,
    updateAppointment,
};
