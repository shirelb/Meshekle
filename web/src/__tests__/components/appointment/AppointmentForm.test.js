import React from 'react';
import {shallow} from "enzyme/build";
import {setupComponent} from "../../testHelpers";
import PageNotFound from "../../../pages/pageNotFound404/PageNotFound";
import store from 'store';
import moment from 'moment';
import {createMemoryHistory} from 'history';
import AppointmentForm from "../../../components/appointment/AppointmentForm";
import serviceProvidersStorage from "../../../storage/serviceProvidersStorage";
import appointmentsStorage from "../../../storage/appointmentsStorage";
import usersStorage from "../../../storage/usersStorage";
import users from "../../jsons/users";
import serviceProviders from "../../jsons/serviceProviders";
import appointmentsOf549963652 from "../../jsons/appointmentsServiceProvider549963652";

const appointmentRequestsOf549963652 = require("../../jsons/appointmentRequestsServiceProvider549963652");

jest.mock("store");
jest.mock("../../../storage/usersStorage");
jest.mock("../../../storage/serviceProvidersStorage");
jest.mock("../../../storage/appointmentsStorage");


describe("AppointmentForm should", () => {
    let wrapper = null;
    let componentInstance = null;
    const pathAdd = '/appointments/set';
    const pathEdit = '/appointments/428/edit';
    let appointmentTest = appointmentsOf549963652[0];
    appointmentTest.clientName = "Padget Creaser";
    let appointmentRequestTest = {
        allDay: false,
        appointmentRequest: appointmentRequestsOf549963652[1],
        end: moment(),
        id: 985,
        start: moment(),
        title: "Catherina Blackmuir"
    };
    appointmentRequestTest.appointmentRequest.optionalTimes = JSON.parse(appointmentRequestTest.appointmentRequest.optionalTimes);
    appointmentRequestTest.appointmentRequest.clientName = "Padget Creaser";
    const historyAdd = createMemoryHistory({
        initialEntries: ['/', '/home', '/appointments', pathAdd],
        initialIndex: 3,
        keyLength: 10,
        getUserConfirmation: null
    });
    const historyEdit = createMemoryHistory({
        initialEntries: ['/', '/home', '/appointments', pathEdit],
        initialIndex: 3,
        keyLength: 10,
        getUserConfirmation: null
    });
    const slotInfo = {
        end: moment(),
        start: moment(),
    };
    const buildProps = (history, path, locationState) => {
        return Object.assign({}, {
            history: history,
            location: {
                pathname: path,
            },
            match: {
                isExact: true,
                path: path,
                url: path,
            },
            serviceProviderRoles: ["appointmentsHairDresser"],
            userOptions: [{
                key: '580624494',
                text: "Padget Creaser",
                value: "Padget Creaser"
            }],
        }, locationState)
    };

    const mockStore = {
        serviceProviderId: "549963652",
        userId: "549963652",
    };

    store.get.mockImplementation((key) => mockStore[key]);
    usersStorage.getUsers.mockResolvedValue(users);
    usersStorage.getUserByUserID.mockImplementation((userId) => Promise.resolve(users.filter(user => user.userId === userId)[0]));
    serviceProvidersStorage.getServiceProviderUserDetails.mockImplementation((serviceProviderId) => Promise.resolve({data: users.filter(user => user.userId === serviceProviderId)[0]}));
    serviceProvidersStorage.getServiceProviderById.mockImplementation((serviceProviderId) => Promise.resolve(serviceProviders.filter(provider =>
        provider.serviceProviderId === serviceProviderId)));
    serviceProvidersStorage.getRolesOfServiceProvider.mockResolvedValue(["appointmentsHairDresser"]);
    serviceProvidersStorage.getServiceProviderById.mockResolvedValue("[\"החלקה\", \"גוונים\", \"צבע\", \"תספורת\"]");
    appointmentsStorage.setAppointment.mockResolvedValue({});
    appointmentsStorage.getServiceProviderAppointments.mockResolvedValue({data: appointmentsOf549963652});
    appointmentsStorage.getServiceProviderAppointmentRequests.mockImplementation(() => Promise.resolve({data: JSON.parse(JSON.stringify(appointmentRequestsOf549963652))}));
    appointmentsStorage.updateAppointment.mockResolvedValue({});
    appointmentsStorage.getAppointmentByAppointmentID.mockResolvedValue({data: appointmentsOf549963652[0]});
    appointmentsStorage.cancelAppointmentById.mockResolvedValue({});

    beforeAll((done) => {
        done();
    });

    afterAll(() => {
    });

    beforeEach((done) => {
        jest.clearAllMocks();
        done();
    });

    afterEach(() => {
    });

    test('match snapshot with slotInfo', async () => {
        const arrResponse = setupComponent('shallow', AppointmentForm, null, buildProps(null, pathAdd, {
            slotInfo: slotInfo,
            submitText: "קבע",
        }), pathAdd);

        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(componentInstance).toMatchSnapshot();
    });

    test('match snapshot with appointment', async () => {
        const arrResponse = setupComponent('shallow', AppointmentForm, null, buildProps(null, pathEdit, {
            appointment: appointmentTest,
            submitText: "עדכן"
        }), pathEdit);
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(componentInstance).toMatchSnapshot();
    });

    test('match snapshot with appointmentRequest', async () => {
        const arrResponse = setupComponent('shallow', AppointmentForm, null, buildProps(null, pathAdd, {
            appointmentRequestEvent: appointmentRequestTest,
            submitText: "קבע",
        }), pathAdd);

        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(componentInstance).toMatchSnapshot();
    });

    test("renders AppointmentForm for /appointments/set", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentForm, historyAdd, buildProps(historyAdd, pathAdd, {
            slotInfo: slotInfo,
            submitText: "קבע",
        }), pathAdd);

        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(AppointmentForm)).toHaveLength(1);
        expect(wrapper.find(PageNotFound)).toHaveLength(0);
    });

    test("renders AppointmentForm for /appointments/428/edit", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentForm, historyEdit, buildProps(historyEdit, pathEdit, {
            appointment: appointmentTest,
            submitText: "עדכן"
        }), pathEdit);
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(AppointmentForm)).toHaveLength(1);
        expect(wrapper.find(PageNotFound)).toHaveLength(0);
    });

    test("mounted with the right data for Add with slotInfo", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentForm, historyAdd, buildProps(historyAdd, pathAdd, {
            slotInfo: slotInfo,
            submitText: "קבע",
        }), pathAdd);

        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(componentInstance.state.appointment).toBeDefined();
        expect(componentInstance.state.appointment.clientName).toEqual("");
        expect(componentInstance.state.appointment.role).toEqual('');
        expect(componentInstance.state.appointment.subject).toEqual([]);
        expect(componentInstance.state.appointment.date).toEqual(moment(slotInfo.start).format('YYYY-MM-DD'));
        expect(componentInstance.state.appointment.startTime).toEqual(moment(slotInfo.start).format("HH:mm"));
        expect(componentInstance.state.appointment.endTime).toEqual(moment(slotInfo.end).format("HH:mm"));
    });

    test("mounted with the right data for Add with appointmentRequestEvent", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentForm, historyAdd, buildProps(historyAdd, pathAdd, {
            appointmentRequestEvent: appointmentRequestTest,
            submitText: "קבע",
        }), pathAdd);

        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(componentInstance.state.appointment).toBeDefined();
        expect(componentInstance.state.appointment.date).toEqual(moment(appointmentRequestTest.start).format("YYYY-MM-DD"));
        expect(componentInstance.state.appointment.startTime).toEqual(moment(appointmentRequestTest.start).format("HH:mm"));
        expect(componentInstance.state.appointment.endTime).toEqual(moment(appointmentRequestTest.end).format("HH:mm"));
        expect(componentInstance.state.appointment.role).toEqual(appointmentRequestTest.appointmentRequest.AppointmentDetail.role);
        expect(componentInstance.state.appointment.subject).toEqual(JSON.parse(appointmentRequestTest.appointmentRequest.AppointmentDetail.subject));
        expect(componentInstance.state.appointment.clientName).toEqual(appointmentRequestTest.appointmentRequest.clientName);
        expect(componentInstance.state.appointment.remarks).toEqual(appointmentRequestTest.appointmentRequest.notes);
    });

    test("mounted with the right data for Edit with appointment", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentForm, historyEdit, buildProps(historyEdit, pathEdit, {
            appointment: appointmentTest,
            submitText: "עדכן"
        }), pathEdit);
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(componentInstance.state.appointment).toBeDefined();
        expect(componentInstance.state.appointment.date).toEqual(moment(appointmentTest.startDateAndTime).format("YYYY-MM-DD"));
        expect(componentInstance.state.appointment.startTime).toEqual(moment(appointmentTest.startDateAndTime).format("HH:mm"));
        expect(componentInstance.state.appointment.endTime).toEqual(moment(appointmentTest.endDateAndTime).format("HH:mm"));
        expect(componentInstance.state.appointment.role).toEqual(appointmentTest.AppointmentDetail.role);
        expect(componentInstance.state.appointment.subject).toEqual(JSON.parse(appointmentTest.AppointmentDetail.subject));
        expect(componentInstance.state.appointment.clientName).toEqual(appointmentTest.clientName);
        expect(componentInstance.state.appointment.remarks).toEqual(appointmentTest.remarks);
    });

    test("handleSubmit with empty fields", async () => {
        const arrResponse = await setupComponent('mount', AppointmentForm, historyAdd, buildProps(historyAdd, pathAdd, {
            slotInfo: slotInfo,
            submitText: "קבע",
        }), pathAdd);

        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        await wrapper.find('AppointmentForm').find('Form').simulate('submit', {
            preventDefault() {
            }
        });
        expect(componentInstance.state.formError).toEqual(true);
    });

    test("handleSubmit in edit mode", async () => {
        const arrResponse = await setupComponent('mount', AppointmentForm, historyEdit, buildProps(historyEdit, pathEdit, {
            appointment: appointmentTest,
            submitText: "עדכן",
            handleSubmit: jest.fn(),
        }), pathEdit);
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        await wrapper.find('AppointmentForm').find('Form').simulate('submit', {
            preventDefault() {
            }
        });
        expect(componentInstance.state.appointment).toEqual({});
    });

    test("handleChange", async () => {
        const handleChangeSpy = jest.spyOn(AppointmentForm.prototype, 'handleChange');

        const arrResponse = await setupComponent('mount', AppointmentForm, historyAdd, buildProps(historyAdd, pathAdd, {
            appointmentRequestEvent: appointmentRequestTest,
            submitText: "קבע",
        }), pathAdd);
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        await wrapper.find('AppointmentForm').find('FormField').at(3).props().onChange({
            target: {
                name: 'remarks',
                value: 'this is the remarks'
            },
        });

        expect(handleChangeSpy).toHaveBeenCalled();
        expect(handleChangeSpy.mock.instances[0].state.appointment.remarks).toEqual('this is the remarks');
    });

    test("handleClear", async () => {
        const arrResponse = await setupComponent('mount', AppointmentForm, historyAdd, buildProps(historyAdd, pathAdd, {
            appointmentRequestEvent: appointmentRequestTest,
            submitText: "קבע",
        }), pathAdd);
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        wrapper.find('AppointmentForm').find('FormButton').at(2).props().onClick({
            preventDefault() {
            }
        });
        expect(componentInstance.state.appointment.date).toEqual('');
    });

    test("render with what the user see for /appointments/set", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentForm, historyAdd, buildProps(historyAdd, pathAdd, {
            slotInfo: slotInfo,
            submitText: "קבע",
        }), pathAdd);

        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const shallowWrapper = wrapper.find('AppointmentForm').dive();

        expect(shallowWrapper.find('Form')).toHaveLength(1);
        expect(shallowWrapper.find('FormField')).toHaveLength(6);
        expect(shallowWrapper.find('FormField').at(0).props().name).toEqual('clientName');
        expect(shallowWrapper.find('FormField').at(0).props().label).toEqual('שם לקוח');
        expect(shallowWrapper.find('FormField').at(0).props().value).toEqual('');
        expect(shallowWrapper.find('FormField').at(1).props().name).toEqual('role');
        expect(shallowWrapper.find('FormField').at(1).props().label).toEqual('ענף');
        expect(shallowWrapper.find('FormField').at(1).props().value).toEqual('');
        expect(shallowWrapper.find('FormField').at(2).props().name).toEqual('remarks');
        expect(shallowWrapper.find('FormField').at(2).props().label).toEqual('הערות');
        expect(shallowWrapper.find('FormField').at(2).props().value).toEqual('');
        expect(shallowWrapper.find('FormField').at(3).props().label).toEqual('תאריך');
        expect(shallowWrapper.find('FormField').at(3).props().value).toEqual(moment(slotInfo.start).format("YYYY-MM-DD"));
        expect(shallowWrapper.find('FormField').at(4).props().label).toEqual('שעת התחלה');
        expect(shallowWrapper.find('FormField').at(4).props().value).toEqual(moment(slotInfo.start).format("HH:mm"));
        expect(shallowWrapper.find('FormField').at(5).props().label).toEqual('שעת סיום');
        expect(shallowWrapper.find('FormField').at(5).props().value).toEqual(moment(slotInfo.end).format("HH:mm"));

        expect(shallowWrapper.find('Message')).toHaveLength(0);
        expect(shallowWrapper.find('Modal')).toHaveLength(0);

        expect(shallowWrapper.find('FormButton')).toHaveLength(3);
        expect(shallowWrapper.find('FormButton').at(0).props().children).toEqual("קבע");
        expect(shallowWrapper.find('FormButton').at(0).props().positive).toBeTruthy();
        expect(shallowWrapper.find('FormButton').at(0).props().type).toEqual("submit");
        expect(shallowWrapper.find('FormButton').at(1).props().children).toEqual("בטל");
        expect(shallowWrapper.find('FormButton').at(1).props().negative).toBeTruthy();
        expect(shallowWrapper.find('FormButton').at(2).props().children).toEqual("נקה הכל");
    });

    test("render with what the user see for /appointments/set with request", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentForm, historyAdd, buildProps(historyAdd, pathAdd, {
            appointmentRequestEvent: appointmentRequestTest,
            submitText: "קבע",
        }), pathAdd);

        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const shallowWrapper = wrapper.find('AppointmentForm').dive();

        expect(shallowWrapper.find('Form')).toHaveLength(1);
        expect(shallowWrapper.find('FormField')).toHaveLength(7);
        expect(shallowWrapper.find('FormField').at(0).props().name).toEqual('clientName');
        expect(shallowWrapper.find('FormField').at(0).props().label).toEqual('שם לקוח');
        expect(shallowWrapper.find('FormField').at(0).props().value).toEqual(appointmentRequestTest.appointmentRequest.clientName);
        expect(shallowWrapper.find('FormField').at(1).props().name).toEqual('role');
        expect(shallowWrapper.find('FormField').at(1).props().label).toEqual('ענף');
        expect(shallowWrapper.find('FormField').at(1).props().value).toEqual(appointmentRequestTest.appointmentRequest.AppointmentDetail.role);
        expect(shallowWrapper.find('FormField').at(2).props().name).toEqual('subject');
        expect(shallowWrapper.find('FormField').at(2).props().label).toEqual('נושא');
        expect(shallowWrapper.find('FormField').at(2).props().value).toEqual(JSON.parse(appointmentRequestTest.appointmentRequest.AppointmentDetail.subject));
        expect(shallowWrapper.find('FormField').at(3).props().name).toEqual('remarks');
        expect(shallowWrapper.find('FormField').at(3).props().label).toEqual('הערות');
        expect(shallowWrapper.find('FormField').at(3).props().value).toEqual(appointmentRequestTest.appointmentRequest.notes);
        expect(shallowWrapper.find('FormField').at(4).props().label).toEqual('תאריך');
        expect(shallowWrapper.find('FormField').at(4).props().value).toEqual(moment(appointmentRequestTest.start).format("YYYY-MM-DD"));
        expect(shallowWrapper.find('FormField').at(5).props().label).toEqual('שעת התחלה');
        expect(shallowWrapper.find('FormField').at(5).props().value).toEqual(moment(appointmentRequestTest.start).format("HH:mm"));
        expect(shallowWrapper.find('FormField').at(6).props().label).toEqual('שעת סיום');
        expect(shallowWrapper.find('FormField').at(6).props().value).toEqual(moment(appointmentRequestTest.end).format("HH:mm"));

        expect(shallowWrapper.find('Message')).toHaveLength(2);
        expect(shallowWrapper.find('Modal')).toHaveLength(1);

        expect(shallowWrapper.find('FormButton')).toHaveLength(3);
        expect(shallowWrapper.find('FormButton').at(0).props().children).toEqual("קבע");
        expect(shallowWrapper.find('FormButton').at(0).props().positive).toBeTruthy();
        expect(shallowWrapper.find('FormButton').at(0).props().type).toEqual("submit");
        expect(shallowWrapper.find('FormButton').at(1).props().children).toEqual("בטל");
        expect(shallowWrapper.find('FormButton').at(1).props().negative).toBeTruthy();
        expect(shallowWrapper.find('FormButton').at(2).props().children).toEqual("נקה הכל");
    });

    test("render with what the user see for /appointments/428/edit", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentForm, historyEdit, buildProps(historyEdit, pathEdit, {
            appointment: appointmentTest,
            submitText: "עדכן"
        }), pathEdit);
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const shallowWrapper = wrapper.find('AppointmentForm').dive();

        expect(shallowWrapper.find('Form')).toHaveLength(1);
        expect(shallowWrapper.find('FormField')).toHaveLength(7);
        expect(shallowWrapper.find('FormField').at(0).props().name).toEqual('clientName');
        expect(shallowWrapper.find('FormField').at(0).props().label).toEqual('שם לקוח');
        expect(shallowWrapper.find('FormField').at(0).props().value).toEqual(appointmentTest.clientName);
        expect(shallowWrapper.find('FormField').at(1).props().name).toEqual('role');
        expect(shallowWrapper.find('FormField').at(1).props().label).toEqual('ענף');
        expect(shallowWrapper.find('FormField').at(1).props().value).toEqual(appointmentTest.AppointmentDetail.role);
        expect(shallowWrapper.find('FormField').at(2).props().name).toEqual('subject');
        expect(shallowWrapper.find('FormField').at(2).props().label).toEqual('נושא');
        expect(shallowWrapper.find('FormField').at(2).props().value).toEqual(JSON.parse(appointmentTest.AppointmentDetail.subject));
        expect(shallowWrapper.find('FormField').at(3).props().name).toEqual('remarks');
        expect(shallowWrapper.find('FormField').at(3).props().label).toEqual('הערות');
        expect(shallowWrapper.find('FormField').at(3).props().value).toEqual(appointmentTest.remarks);
        expect(shallowWrapper.find('FormField').at(4).props().label).toEqual('תאריך');
        expect(shallowWrapper.find('FormField').at(4).props().value).toEqual(moment(appointmentTest.startDateAndTime).format("YYYY-MM-DD"));
        expect(shallowWrapper.find('FormField').at(5).props().label).toEqual('שעת התחלה');
        expect(shallowWrapper.find('FormField').at(5).props().value).toEqual(moment(appointmentTest.startDateAndTime).format("HH:mm"));
        expect(shallowWrapper.find('FormField').at(6).props().label).toEqual('שעת סיום');
        expect(shallowWrapper.find('FormField').at(6).props().value).toEqual(moment(appointmentTest.endDateAndTime).format("HH:mm"));

        expect(shallowWrapper.find('Message')).toHaveLength(0);
        expect(shallowWrapper.find('Modal')).toHaveLength(0);

        expect(shallowWrapper.find('FormButton')).toHaveLength(3);
        expect(shallowWrapper.find('FormButton').at(0).props().children).toEqual("עדכן");
        expect(shallowWrapper.find('FormButton').at(0).props().positive).toBeTruthy();
        expect(shallowWrapper.find('FormButton').at(0).props().type).toEqual("submit");
        expect(shallowWrapper.find('FormButton').at(1).props().children).toEqual("בטל");
        expect(shallowWrapper.find('FormButton').at(1).props().negative).toBeTruthy();
        expect(shallowWrapper.find('FormButton').at(2).props().children).toEqual("נקה הכל");
    });
});
