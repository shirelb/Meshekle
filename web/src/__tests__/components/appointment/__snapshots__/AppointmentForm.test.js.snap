// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentForm should match snapshot with appointment 1`] = `undefined`;

exports[`AppointmentForm should match snapshot with appointmentRequest 1`] = `undefined`;

exports[`AppointmentForm should match snapshot with slotInfo 1`] = `undefined`;
